### ISSUE TITLE
FTPClient errors should provide ftp errorcode

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3490]
|<img alt="tarjei's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| tarjei reported|
|-|-|
|Trac ID|trac#3490|
|Type|defect|
|Created|2008-10-23 10:48:52Z|
|Branch|https://github.com/twisted/twisted/tree/ftp-commandfailed-docs-3490|

This one is subtle. 

The problem starts in ftp.FTPClient.changeDirectory():

It parses the result of a mkd command to check if the errorcode is 250. 

There is however an errorcode that is ambigous:
550 : Directory exists

What if you just want to ensure the directory exists? Then this is a success. If not, it's a failure. 

How should this be reflected in what is returned in the result? I'm not sure about that part, and therefore I have not tried to code up a suggested patch, but I think the best way might be to probvide the errorcode with the result somehow. Please advise.

<details><summary>Searchable metadata</summary>

```
trac-id__3490 3490
type__defect defect
reporter__tarjei tarjei
priority__normal normal
milestone__ 
branch__branches_ftp_commandfailed_docs_3490 branches/ftp-commandfailed-docs-3490
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__ftp ftp
keywords__ftp__client ftp, client
time__1224758932000000 1224758932000000
changetime__1227017650000000 1227017650000000
version__None None
owner__ 
cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

If you get a 550, then doesn't `changeDirectory` return a Deferred which fails with `CommandFailed` with the entire result as its `.args[0]`?  Then you can parse the result and see it was a 550.

**Comment by twisted-trac** [comment]
|<img alt="tarjei's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|tarjei commented|
|-|-|

Good point. What about adding this to the docs for FtpClient:

--- twisted/protocols/ftp.py	(revision 24235)
+++ twisted/protocols/ftp.py	(working copy)
@@ -2181,6 +2181,13 @@
 
     Supports active and passive transfers.
 
+
+    A note on error checks. On failure, a CommandFailed command is
+    returned. To check the errorvalue, inspect result.value.args[0]
+    to find the relevant errorcode.
+
+    You can find descriptions of the different errorcodes in RFC 959.
+
     @ivar passive: See description in __init__.
     """
     connectFactory = reactor.connectTCP

**Comment by twisted-trac** [comment]
|<img alt="tarjei's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|tarjei commented|
|-|-|

Args, the above should be:

```
--- twisted/protocols/ftp.py  (revision 24235)
+++ twisted/protocols/ftp.py  (working copy)
@@ -2181,6 +2181,13 @@
 
     Supports active and passive transfers.
 
+
+    A note on error checks. On failure, a CommandFailed command is
+    returned. To check the errorvalue, inspect result.value.args[0]
+    to find the relevant errorcode.
+
+    You can find descriptions of the different errorcodes in RFC 959.
+
     @ivar passive: See description in __init__.
     """ 
     connectFactory = reactor.connectTCP
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

Something like this is a good idea.  I'll come back to this ticket after I review the other FTP tickets currently up for review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [25406]) Branching to 'ftp-commandfailed-docs-3490'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve set owner to @exarkun|
|-|-|

+1.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set status to `closed`|
|-|-|

(In [25413]) Merge ftp-commandfailed-docs-3490

Author: exarkun
Reviewer: therve
Fixes: [#3490](https://github.com/twisted/twisted/issues/3490)

Clarify the API documentation for `FTPClient` with regard to the results and
failure results of methods which expose protocol actions.  Specifically, explain
how error response codes are returned and where to learn more about them.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
